keps/sig-scheduling/5710-workload-aware-preemption/README.md

```diff
@@ -0,0 +1,1165 @@
+
+#### Preemption of Multihost Inference
+
+When running multihost inference using LeaderWorkerSet, I want to ensure that its single
+replica (one leader and N workers) will not be partially preempted. If at least one of
+such pods is not running, the others are not able to serve and are just wasting resources
+in the cluster.
+
+#### Preemption of Multihost Inference that can run in a degraded mode
+
+When running multihost inference using LeaderWorkerSet that can run in a degraded mode,
+I don't want to preempt the whole replica (one leader and N workers) if a single worker
+would be preempted, because I prefer to serve in a degraded mode than be completely
+disrupted.
+
+#### Preemption cost
+
+When a long-running AI Training job, its cost of preemption differs over time and depends
+on how long ago it checkpointed its state. I want to be able to somehow influence the
+preemption priority of my job based on how much would it really cost me.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Preemption unit
+
+We start with defining what is a unit of preemption. While we can imagine usecases with preemption
+unit being an arbitrary group of pods, in majority of real world usecases this is actually aligned
+with the scheduling unit. In other words, the group of pods that should be preempted together matches
+a group that was initially scheduled together as a gang.
+
+Trying to formalize it, we define `WorkloadPortion` as one of {all pods in a PodGroup replica or
+a single pod}. With that definition both scheduling unit and preemption units can only be
+`WorkloadPortions`.
+
+In the future, we may want to support usecases when a single scheduling unit consists of multiple
+preemption groups, but we leave that usecase as a future extension (it can be addressed when we
+decide to extend Workload API with PodSubGroup concept - for more details see
+[API Design For Gang and Workload-Aware Scheduling]). However, we never expect preemption unit to
+be larger than scheduling unit.
+
+Based on that, we will extend the the existing `GangSchedulingPolicy` as following:
+
+```golang
+// PreemptionMode describes the mode in which a PodGroup can be preempted.
+// +enum
+type PreemptionMode string
+
+const (
+  // PreemptionModePod means that individual pods can be preempted independently.
+  PreemptionModePod = "Pod"
```

Comment:
When gang scheduling policy is configured with `PreemptionModePod`, is it correct to assume that pods are evicted individually until the PodGroup size drops below `minCount`, at which point the remaining pods are preempted collectively?

Reply:
I don't think that's the case. I expect that those actually do not care about minCount and they are either allowing pod by pod preemption or preemption of the whole pod group no matter what the size of pod group is in relation to its minCount. Maybe the behavior you describe can be added as a third option? 
 
Btw, there is an ongoing discussion about exact minCount semantics in workload scheduling doc https://github.com/kubernetes/enhancements/pull/5730#discussion_r2608747335

Reply:
Right - I expanded the comment trying to clarify this.

Reply:
It does seem like there's a middle ground between "any pod can be preempted, ignoring the atomicity of the pod group" and "preempt the whole podgroup". In fact it would seem pretty common that you might need to preempt a few pods from one group to make space (again thinking topological constraints) for a new workload, and that those pods from that podgroup would still need a place to start back up. In other words, the podgroup still needs to be able to be made complete, otherwise the cost metric should be "whole podgroup lost".